javascript-typing-tutor/main.js

```diff
@@ -0,0 +1,82 @@
+function NewChallenge(string) {
+  var challenge = {};
+  challenge.complete = false;
+  challenge.text = string;
+  challenge.spans = [];
+  challenge.charErrors = 0;
+  challenge.time = 0;
+  challenge.accuracy = 0;
+  challenge.started = false;
+  challenge.errors = 0;
+  challenge.warned = false;
+  challenge.div = document.querySelector('.challenge-container');
+  document.body.appendChild(challenge.div);
+  for (var i = 0; i < string.length; i++) {
+    var $span = document.createElement('span');
+    $span.textContent = string[i];
+    challenge.spans.push($span);
+    challenge.div.appendChild($span);
+  }
+  challenge.index = 0;
+  return challenge;
+}
+var challenge = NewChallenge('The quick brown fox jumps over the lazy dog.');
+document.addEventListener('keydown', logKey);
+function logKey(event) {
+  if (event.key.length > 1 || challenge.complete) {
+    return;
+  }
+  if (event.key === challenge.text[challenge.index]) {
+    challenge.spans[challenge.index].className = 'typed';
+    if (!challenge.started) {
+      challenge.started = true;
+      challenge.time = Date.now();
+    }
+    if (challenge.charErrors > 0) {
+      challenge.errors++;
+      challenge.charErrors = 0;
+    }
+    if (challenge.spans[challenge.index + 1] !== undefined) {
+      challenge.spans[challenge.index + 1].className = 'cursor';
+    }
+    challenge.index++;
+    if (challenge.index > challenge.text.length - 1) {
+      challenge.complete = true;
+      challenge.time = Date.now() - challenge.time;
+      score();
+    }
+  } else {
+    challenge.charErrors++;
+    if (challenge.charErrors <= 1) {
+      challenge.spans[challenge.index].className = 'cursor error';
+    } else if (challenge.charErrors < 2) {
+      challenge.spans[challenge.index].className = 'cursor big error';
+    } else if (challenge.charErrors < 4) {
+      challenge.spans[challenge.index].className = 'cursor chunky error';
+    } else if (challenge.charErrors < 6) {
+      challenge.spans[challenge.index].className = 'cursor huge error';
+    }
+  }
+}
+
+function score() {
+  challenge.wpm = Math.round(challenge.text.length / (challenge.time / 1000) * 60 / 5);
+  var $reset = document.createElement('a');
+  $reset.className = '.reset-button';
+  $reset.href = 'index.html';
```

Comment:
Oh dear, reloading a page entirely is a very costly way to restart something. Loading a document is _extremely_ taxing for the browser. Instead, you should reset the value in your data model to match what they would be when the page loaded, and update the DOM to reflect that as well.